File: backend/shared/nodejs/encryption/examples/lambda-handler-example.js
/**
 * Example Lambda handler demonstrating encryption utilities usage
 */

const { FieldEncryption, KMSEncryption } = require('../index');

// Initialize encryption services (reuse across invocations)
const fieldEncryption = new FieldEncryption();
const kmsEncryption = new KMSEncryption();

/**
 * Lambda handler to store encrypted patient records
 *
 * Example event:
 * {
 *   "body": "{\"patient_name\": \"Rajesh Kumar\", \"age\": 45, ...}",
 *   "pathParameters": {"patientId": "123"}
 * }
 */
exports.storePatientRecord = async (event, context) => {
  try {
    // Parse request
    const patientRecord = JSON.parse(event.body);
    const patientId = event.pathParameters.patientId;

    // Encrypt sensitive fields
    const encryptedRecord = await fieldEncryption.encryptRecord(patientRecord, patientId, {
      operation: 'store_record',
      source: 'api_gateway',
    });

    // Add metadata
    encryptedRecord.patient_id = patientId;
    encryptedRecord.encrypted_at = context.awsRequestId;

    // Store in DynamoDB (pseudo-code)
    // await dynamodb.putItem({
    //   TableName: 'Patients',
    //   Item: encryptedRecord
    // });

    return {
      statusCode: 200,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      },
      body: JSON.stringify({
        message: 'Patient record stored successfully',
        patient_id: patientId,
      }),
    };
  } catch (error) {
    console.error('Error storing patient record:', error);
    return {
      statusCode: error.message.includes('Invalid') ? 400 : 500,
      body: JSON.stringify({
        error: error.message || 'Internal server error',
      }),
    };
  }
};

/**
 * Lambda handler to retrieve and decrypt patient records
 *
 * Example event:
 * {
 *   "pathParameters": {"patientId": "123"},
 *   "queryStringParameters": {"fields": "patient_name,age"}
 * }
 */
exports.retrievePatientRecord = async (event) => {
  try {
    const patientId = event.pathParameters.patientId;

    // Get requested fields (optional partial decryption)
    const queryParams = event.queryStringParameters || {};
    let requestedFields = null;
    if (queryParams.fields) {
      requestedFields = queryParams.fields.split(',');
    }

    // Retrieve from DynamoDB (pseudo-code)
    // const result = await dynamodb.getItem({
    //   TableName: 'Patients',
    //   Key: { patient_id: patientId }
    // });
    // const encryptedRecord = result.Item;

    // Mock encrypted record for example
    const encryptedRecord = {
      patient_id: patientId,
      patient_name: 'AQICAHh...encrypted...',
      patient_name_encrypted: true,
      age: 45,
      medical_history: 'AQICAHh...encrypted...',
      medical_history_encrypted: true,
    };

    // Decrypt record (partial or full)
    const decryptedRecord = await fieldEncryption.decryptRecord(
      encryptedRecord,
      patientId,
      null,
      requestedFields
    );

    return {
      statusCode: 200,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      },
      body: JSON.stringify(decryptedRecord),
    };
  } catch (error) {
    console.error('Error retrieving patient record:', error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: 'Internal server error' }),
    };
  }
};

/**
 * Lambda handler to encrypt medical document metadata
 * Uses direct KMS encryption for individual fields
 */
exports.encryptMedicalDocument = async (event) => {
  try {
    const documentMetadata = JSON.parse(event.body);
    const patientId = documentMetadata.patient_id;

    // Encrypt specific sensitive fields using KMS directly
    const encryptionContext = {
      patient_id: patientId,
      service: 'vaidyalink',
      data_type: 'document_metadata',
    };

    const encryptedMetadata = {
      document_id: documentMetadata.document_id,
      patient_id: patientId,
      document_type: documentMetadata.document_type,
      encrypted_notes: await kmsEncryption.encrypt(
        documentMetadata.doctor_notes,
        encryptionContext
      ),
      encrypted_diagnosis: await kmsEncryption.encrypt(
        documentMetadata.diagnosis,
        encryptionContext
      ),
      upload_date: documentMetadata.upload_date,
    };

    // Store metadata
    // await dynamodb.putItem({
    //   TableName: 'Documents',
    //   Item: encryptedMetadata
    // });

    return {
      statusCode: 200,
      body: JSON.stringify({
        message: 'Document metadata encrypted and stored',
        document_id: documentMetadata.document_id,
      }),
    };
  } catch (error) {
    console.error('Error encrypting document:', error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: 'Internal server error' }),
    };
  }
};

/**
 * Lambda handler to re-encrypt data with new KMS key
 * Used during key rotation
 */
exports.rotateEncryptionKeys = async (event) => {
  try {
    // Get records to re-encrypt
    const patientId = event.pathParameters.patientId;
    const newKeyId = process.env.NEW_KMS_KEY_ID;

    if (!newKeyId) {
      throw new Error('NEW_KMS_KEY_ID not configured');
    }

    // Retrieve encrypted record
    // const result = await dynamodb.getItem(...);
    // const encryptedRecord = result.Item;

    // Re-encrypt each encrypted field
    const oldContext = {
      patient_id: patientId,
      service: 'vaidyalink',
      data_type: 'phi',
    };

    const newContext = {
      ...oldContext,
      key_version: 'v2',
    };

    // Re-encrypt sensitive fields
    // for (const field of encryptedFields) {
    //   const oldCiphertext = encryptedRecord[field];
    //   const newCiphertext = await kmsEncryption.reEncrypt(
    //     oldCiphertext,
    //     newKeyId,
    //     oldContext,
    //     newContext
    //   );
    //   encryptedRecord[field] = newCiphertext;
    // }

    // Update record
    // await dynamodb.putItem(...);

    return {
      statusCode: 200,
      body: JSON.stringify({
        message: 'Keys rotated successfully',
        patient_id: patientId,
      }),
    };
  } catch (error) {
    console.error('Error rotating keys:', error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: 'Internal server error' }),
    };
  }
};

/**
 * Example: Batch encryption for multiple records
 */
exports.batchEncryptRecords = async (event) => {
  try {
    const records = JSON.parse(event.body).records;

    const encryptedRecords = await Promise.all(
      records.map(async (record) => {
        return await fieldEncryption.encryptRecord(record, record.patient_id);
      })
    );

    // Store batch
    // await dynamodb.batchWriteItem(...);

    return {
      statusCode: 200,
      body: JSON.stringify({
        message: `${encryptedRecords.length} records encrypted successfully`,
      }),
    };
  } catch (error) {
    console.error('Error batch encrypting:', error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: 'Internal server error' }),
    };
  }
};

/**
 * Example: Generate data key for envelope encryption
 */
exports.generateDataKeyForFile = async (event) => {
  try {
    const { patient_id, file_type } = JSON.parse(event.body);

    // Generate data key for encrypting large file
    const dataKey = await kmsEncryption.generateDataKey('AES_256', {
      patient_id,
      file_type,
      service: 'vaidyalink',
    });

    // Return encrypted data key (store with file)
    // Use plaintext key to encrypt file locally
    // Securely delete plaintext key after use

    return {
      statusCode: 200,
      body: JSON.stringify({
        encrypted_data_key: dataKey.ciphertext.toString('base64'),
        // Never return plaintext key in response!
      }),
    };
  } catch (error) {
    console.error('Error generating data key:', error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: 'Internal server error' }),
    };
  }
};

File: backend/shared/python/encryption/examples/lambda_handler_example.py
"""
Example Lambda handler demonstrating encryption utilities usage
"""

import json
import os
from typing import Dict, Any
from encryption import FieldEncryption, KMSEncryption


# Initialize encryption services (reuse across invocations)
field_encryption = FieldEncryption()
kms_encryption = KMSEncryption()


def store_patient_record(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Lambda handler to store encrypted patient records

    Example event:
    {
        "body": "{\"patient_name\": \"Rajesh Kumar\", \"age\": 45, ...}",
        "pathParameters": {"patientId": "123"}
    }
    """
    try:
        # Parse request
        patient_record = json.loads(event['body'])
        patient_id = event['pathParameters']['patientId']

        # Encrypt sensitive fields
        encrypted_record = field_encryption.encrypt_record(
            patient_record,
            patient_id=patient_id,
            additional_context={
                'operation': 'store_record',
                'source': 'api_gateway'
            }
        )

        # Add metadata
        encrypted_record['patient_id'] = patient_id
        encrypted_record['encrypted_at'] = context.aws_request_id

        # Store in DynamoDB (pseudo-code)
        # dynamodb.put_item(TableName='Patients', Item=encrypted_record)

        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({
                'message': 'Patient record stored successfully',
                'patient_id': patient_id
            })
        }

    except ValueError as e:
        return {
            'statusCode': 400,
            'body': json.dumps({'error': f'Invalid request: {str(e)}'})
        }
    except Exception as e:
        print(f"Error storing patient record: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({'error': 'Internal server error'})
        }


def retrieve_patient_record(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Lambda handler to retrieve and decrypt patient records

    Example event:
    {
        "pathParameters": {"patientId": "123"},
        "queryStringParameters": {"fields": "patient_name,age"}
    }
    """
    try:
        patient_id = event['pathParameters']['patientId']

        # Get requested fields (optional partial decryption)
        query_params = event.get('queryStringParameters', {})
        requested_fields = None
        if query_params and 'fields' in query_params:
            requested_fields = query_params['fields'].split(',')

        # Retrieve from DynamoDB (pseudo-code)
        # encrypted_record = dynamodb.get_item(
        #     TableName='Patients',
        #     Key={'patient_id': patient_id}
        # )['Item']

        # Mock encrypted record for example
        encrypted_record = {
            'patient_id': patient_id,
            'patient_name': 'AQICAHh...encrypted...',
            'patient_name_encrypted': True,
            'age': 45,
            'medical_history': 'AQICAHh...encrypted...',
            'medical_history_encrypted': True
        }

        # Decrypt record (partial or full)
        decrypted_record = field_encryption.decrypt_record(
            encrypted_record,
            patient_id=patient_id,
            fields_to_decrypt=requested_fields
        )

        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps(decrypted_record)
        }

    except Exception as e:
        print(f"Error retrieving patient record: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({'error': 'Internal server error'})
        }


def encrypt_medical_document(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Lambda handler to encrypt medical document metadata
    Uses direct KMS encryption for individual fields
    """
    try:
        document_metadata = json.loads(event['body'])
        patient_id = document_metadata['patient_id']

        # Encrypt specific sensitive fields using KMS directly
        encryption_context = {
            'patient_id': patient_id,
            'service': 'vaidyalink',
            'data_type': 'document_metadata'
        }

        encrypted_metadata = {
            'document_id': document_metadata['document_id'],
            'patient_id': patient_id,
            'document_type': document_metadata['document_type'],
            'encrypted_notes': kms_encryption.encrypt(
                document_metadata['doctor_notes'],
                encryption_context
            ),
            'encrypted_diagnosis': kms_encryption.encrypt(
                document_metadata['diagnosis'],
                encryption_context
            ),
            'upload_date': document_metadata['upload_date']
        }

        # Store metadata
        # dynamodb.put_item(TableName='Documents', Item=encrypted_metadata)

        return {
            'statusCode': 200,
            'body': json.dumps({
                'message': 'Document metadata encrypted and stored',
                'document_id': document_metadata['document_id']
            })
        }

    except Exception as e:
        print(f"Error encrypting document: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({'error': 'Internal server error'})
        }


def rotate_encryption_keys(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Lambda handler to re-encrypt data with new KMS key
    Used during key rotation
    """
    try:
        # Get records to re-encrypt
        patient_id = event['pathParameters']['patientId']
        new_key_id = os.environ.get('NEW_KMS_KEY_ID')

        if not new_key_id:
            raise ValueError('NEW_KMS_KEY_ID not configured')

        # Retrieve encrypted record
        # encrypted_record = dynamodb.get_item(...)

        # Re-encrypt each encrypted field
        old_context = {
            'patient_id': patient_id,
            'service': 'vaidyalink',
            'data_type': 'phi'
        }

        new_context = old_context.copy()
        new_context['key_version'] = 'v2'

        # Re-encrypt sensitive fields
        # for field in encrypted_fields:
        #     old_ciphertext = encrypted_record[field]
        #     new_ciphertext = kms_encryption.re_encrypt(
        #         old_ciphertext,
        #         new_key_id,
        #         old_context,
        #         new_context
        #     )
        #     encrypted_record[field] = new_ciphertext

        # Update record
        # dynamodb.put_item(...)

        return {
            'statusCode': 200,
            'body': json.dumps({
                'message': 'Keys rotated successfully',
                'patient_id': patient_id
            })
        }

    except Exception as e:
        print(f"Error rotating keys: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({'error': 'Internal server error'})
        }
